sihl/src/web_migration.ml

```diff
@@ -0,0 +1,62 @@
+let log_src = Logs.Src.create "sihl.middleware.migration"
+
+module Logs = (val Logs.src_log log_src : Logs.LOG)
+
+exception Exception
+
+let template migrations =
+  let migrations = migrations |> List.map fst |> String.concat ", " in
+  Format.asprintf
+    {|
+<!doctype html>
+<html lang="en">
+
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+  <title>Pending migrations</title>
+  <style>
+    %s
+  </style>
+</head>
+
+<body class="antialiased">
+  <div class="py-4 sm:py-12">
+    <div class="max-w-8xl mx-auto px-4 sm:px-6 lg:px-8">
+      <h2 class="text-2xl leading-8 font-semibold font-display text-gray-900 sm:text-3xl sm:leading-9">
+        You have pending migrations: %s
+      </h2>
+      <div class="mt-0 mb-4 text-gray-600">
+        Run `make sihl migrate` to apply the pending migrations.
+      </div>
+      <p>
+    </div>
+  </div>
+</body>
+
+</html>
+    |}
+    Web_error.style
+    migrations
+;;
+
+let middleware pending_migrations =
+  let filter handler req =
+    let%lwt migrations = pending_migrations () in
+    if List.length migrations > 0
+    then (
+      Logs.err (fun m ->
+          m "There are %d pending migrations" (List.length migrations));
+      Logs.info (fun m ->
+          m "Run 'sihl migrate' to apply the pending migrations");
+      if Core_configuration.is_production ()
+         (* The error handler shows internal server error 500 *)
+      then raise Exception
```

Comment:
Are pending migrations this severe? Maybe we shouldn't even allow starting the app if there are pending migrations?

Reply:
That is actually a good question. Maybe the HTTP server shouldn't even start with pending migrations. The repositories assumes a different schema than there is and you have to expect runtime errors. What do you think?

I took this middleware idea from Rails, but with Sihl we have actual service lifecycles that we can use. :)

Reply:
After thinking about it, I don't think we should restrict a Sihl app that much. Let's just print a warning in production for every request and fail in development using the middleware approach.

Reply:
Ok, I don't know what the best approach is, but isn't it dangerous allowing a prod app to start with missing migrations? I propose this: By default the app doesn't even start with missing migrations in either dev and prod. As a sihl user, you can disable that block with an env var "ALLOW_MISSING_MIGRATIONS" or similar. What do you think?

Reply:
It is more complex to define what it means for a Sihl app to start. You can run all sorts of commands involving all sorts of subsets of services. From what I read Rails allows you to start production with pending migrations, that is why we should allow it.

Reply:
Could we not say that anything requiring the database cannot start?

Reply:
We could say that the migration service doesn't start if there are pending migrations, but that seems to be too strict.

Reply:
Ok, then let's leave it like you proposed. => Merging